lib/SPIRV/SPIRVReader.cpp

```diff
@@ -3555,12 +3555,79 @@ void SPIRVToLLVM::transGlobalAnnotations() {
   }
 }
 
+static llvm::MDNode *
+transDecorationsToMetadataList(llvm::LLVMContext *Context,
+                               std::vector<SPIRVDecorate const *> Decorates) {
+  SmallVector<Metadata *, 4> MDs;
+  MDs.reserve(Decorates.size());
+  for (const auto *Deco : Decorates) {
+    std::vector<Metadata *> OPs;
+    auto *KindMD = ConstantAsMetadata::get(
+        ConstantInt::get(Type::getInt32Ty(*Context), Deco->getDecorateKind()));
+    OPs.push_back(KindMD);
+    switch (static_cast<size_t>(Deco->getDecorateKind())) {
+    case DecorationLinkageAttributes: {
+      const auto *const LinkAttrDeco =
+          static_cast<const SPIRVDecorateLinkageAttr *>(Deco);
+      auto *const LinkNameMD =
+          MDString::get(*Context, LinkAttrDeco->getLinkageName());
+      auto *const LinkTypeMD = ConstantAsMetadata::get(ConstantInt::get(
+          Type::getInt32Ty(*Context), LinkAttrDeco->getLinkageType()));
+      OPs.push_back(LinkNameMD);
+      OPs.push_back(LinkTypeMD);
+      break;
+    }
+    case DecorationMergeINTEL: {
+      const auto MergeAttrLits = Deco->getVecLiteral();
+      std::string FirstString = getString(MergeAttrLits);
+      std::string SecondString =
+          getString(MergeAttrLits.cbegin() + getVec(FirstString).size(),
+                    MergeAttrLits.cend());
+      OPs.push_back(MDString::get(*Context, FirstString));
+      OPs.push_back(MDString::get(*Context, SecondString));
+      break;
+    }
+    case DecorationMemoryINTEL:
+    case DecorationUserSemantic: {
+      auto *const StrMD =
+          MDString::get(*Context, getString(Deco->getVecLiteral()));
+      OPs.push_back(StrMD);
+      break;
+    }
+    default: {
+      for (const SPIRVWord Lit : Deco->getVecLiteral()) {
+        auto *const LitMD = ConstantAsMetadata::get(
+            ConstantInt::get(Type::getInt32Ty(*Context), Lit));
+        OPs.push_back(LitMD);
+      }
+      break;
+    }
+    }
+    MDs.push_back(MDNode::get(*Context, OPs));
+  }
+  return MDNode::get(*Context, MDs);
+}
+
+void SPIRVToLLVM::transVarDecorationsToMetadata(SPIRVValue *BV, Value *V) {
+  if (!BV->isVariable())
+    return;
+
+  if (auto *GV = dyn_cast<GlobalVariable>(V)) {
+    std::vector<SPIRVDecorate const *> Decorates = BV->getDecorations();
+    if (!Decorates.empty()) {
+      MDNode *MDList = transDecorationsToMetadataList(Context, Decorates);
```

Comment:
It does look like, that in the end after reverse translation we will have both decoration metadata and it's previously defined representation in LLVM IR module, correct? If so, it kind of looking incorrect, since SPIR-V friendly LLVM IR should be only emitted under SPV-IR flag (and decoration metadata is SPIR-V friendly form). But at the same time, I'm 100% sure, that none of SPIR-V centric LLVM IR device backends are ready to switch from usual representation to decoration metadata representation, so I believe, we should leave this code as is, but with some comment added, WDYT?

Reply:
If I'm correct about duality of LLVM IR output, could you please reflect it in the tests?

Reply:
You're absolutely right! It didn't seem like a lot of extra work to only generate the metadata if in SPIR-V friendly mode so I went with that instead. The tests have also been changed to check that the metadata is only there if SPV-IR mode is set. Is this okay or would you prefer I revert it and just add a comment?

Reply:
I'd gather opinions from @aratajew @svenvh and @AlexeySotkin 

Folks, like this in SPIR-V friendly IR after the translation !spirv.(Param)Decoration will appear and will replace previous representation of some features, for example kernel_arg_type_quals carrying volatile/restrict qualifiers will replaced with the mentioned SPIR-V friendly form. Even if I believe, that it's the correct way to go, I'm afraid, that right now it might break gathering of Arg type qualifiers in OpenCL runtime, since !spirv.(Param)Decoration metadata aren't supported in any devices' compilers yet. WDYT?

Reply:
In my opinion, `!spirv.(Param)Decoration` metadata representation should be used only when a user triggers translation with `--spirv-target-env=SPV-IR` parameter. When `--spirv-target-env=CL2.0` or `--spirv-target-env=CL1.2` is specified, then `SPIRVToOCL` pass should translate `!spirv.(Param)Decoration` to OpenCL representation (`kernel_arg_type_quals` etc.). As I can see, this PR doesn't include any changes in `SPIRVToOCL` pass, so I agree with @MrSidims that this change will break  OpenCL backends that are based `CL2.0` or `CL1.2` translation.

Also, just out of curiousity, @steffenlarsen what is the main motivation to use `!spirv.(Param)Decoration` for SPIR-V Friendly IR path instead of OpenCL representation? Is it because not all SPIRV decorations have their's corresponding OpenCL metadata?

Reply:
> In my opinion, !spirv.(Param)Decoration metadata representation should be used only when a user triggers translation with --spirv-target-env=SPV-IR parameter. When --spirv-target-env=CL2.0 or --spirv-target-env=CL1.2 is specified, then SPIRVToOCL pass should translate !spirv.(Param)Decoration to OpenCL representation (kernel_arg_type_quals etc.). As I can see, this PR doesn't include any changes in SPIRVToOCL pass, so I agree with @MrSidims that this change will break OpenCL backends that are based CL2.0 or CL1.2 translation.

I see that I misunderstood the concern. This change properly produces metadata based on `--spirv-target`, but the concern is whether OpenCL Runtimes based on `SPV-IR` translation path will be able to consume new metadata representation. In case of Intel OpenCL GPU backend, the compiler is in the middle of transition from internal SPIRVReader to Khronos one, so there will be a need to implement a support for new metadata.

Reply:
> Also, just out of curiousity, @steffenlarsen what is the main motivation to use !spirv.(Param)Decoration for SPIR-V Friendly IR path instead of OpenCL representation? Is it because not all SPIRV decorations have their's corresponding OpenCL metadata?

Yes, that is definitely a driving factor. `!spirv.(Param)Decoration` offers more freedom at the cost of making it more SPIR-V specific. In principle any implementation could implement some extension decorator that the translator does not yet support and still be able to produce SPIR-V using that decorator by generating `!spirv.(Param)Decoration`, assuming the parameter space is simple. Additionally, I personally think there is value to the generality of `!spirv.(Param)Decoration`, i.e. the parsing of the metadata trees are overlapping for the two current variants and future versions would likely have the same structure.